Modules/@babylonjs/react-native/shared/BabylonNative.cpp

```diff
@@ -0,0 +1,230 @@
+                {
+                    m_resolveInitPromise = [&rt, resolve{ std::make_shared<jsi::Value>(rt, args[0]) }]()
+                    {
+                        resolve->asObject(rt).asFunction(rt).call(rt);
+                    };
+                    return {};
+                })
+            );
+
+            // Initialize Babylon Native core components
+            JsRuntime::CreateForJavaScript(m_env, CreateJsRuntimeDispatcher(m_env, jsiRuntime, m_jsCallInvoker, m_isRunning));
+
+            // Initialize Babylon Native plugins
+            Plugins::NativeXr::Initialize(m_env);
+            m_nativeInput = &Babylon::Plugins::NativeInput::CreateForJavaScript(m_env);
+            
+            // Initialize Babylon Native polyfills
+            Polyfills::Window::Initialize(m_env);
+
+            // NOTE: React Native's XMLHttpRequest is slow and allocates a lot of memory. This does not override
+            // React Native's implementation, but rather adds a second one scoped to Babylon and used by WebRequest.ts.
+            Polyfills::XMLHttpRequest::Initialize(m_env);
+        }
+
+        ~ReactNativeModule() override
+        {
+            *m_isRunning = false;
+            Napi::Detach(m_env);
+        }
+
+        // NOTE: This only happens when the JS engine is shutting down (other than when the app exits, this only
+        //       happens during a dev mode reload). In this case, EngineHook.ts won't call NativeEngine.dispose,
+        //       so we need to manually do it here to properly clean up these resources.
+        void Deinitialize()
+        {
+            if (m_disposeEngine)
+            {
+                m_disposeEngine();
+                m_disposeEngine = {};
+            }
+        }
+
+        void UpdateView(void* windowPtr, size_t width, size_t height)
+        {
+            // TODO: We shouldn't have to dispatch to the JS thread most of this, but not doing so results in a crash.
+            //       I don't understand the issue yet, but for now just retain the pre-refactor logic. We'll need to
+            //       resolve this to enable manual non-JS thread rendering. Note this only repros in release builds
+            //       where we actually call ResetView.
+            m_jsCallInvoker->invokeAsync([this, windowPtr, width, height]()
+            {
+                if (!m_graphics)
+                {
+                    m_graphics = Graphics::CreateGraphics(windowPtr, width, height);
+                    m_graphics->AddToJavaScript(m_env);
+                    Plugins::NativeEngine::Initialize(m_env, true);
+                    m_resolveInitPromise();
+                }
+                else
+                {
+                    m_graphics->UpdateWindow(windowPtr);
+                    m_graphics->UpdateSize(width, height);
+                    m_graphics->EnableRendering();
+                }
```

Comment:
for UWP we will need a mechanism to allow for calling graphics render functionality from the UI thread. If the thought is that we will have a different BabylonNative setup for UWP, that is fine too

Reply:
My intent is that this will be shared amongst all platforms, including UWP. I understand that the platform specific code needs to be able to determine the rendering thread and probably manually call `RenderCurrentFrame` directly, but that was just too much to try to fit into this already disruptive change. My plan was to tackle that separately after this PR.